### PR TITLE
208/rr add any number of extra fields

### PR DESCRIPTION
**Pull request by mischa-s**
involves some commenting of tests that aren't replaced because they are waiting on https://govtnz.atlassian.net/jira/software/projects/RR/boards/81/backlog?issueParent=26557%2C26136&selectedIssue=RR-263 as per the direction of the product owner

### CODE REVIEW

**Comment by mischa-s** [comment]
travis is failing because capybara isn't able to see React, have to think about how test that react has rendered

File: app/javascript/components/EditForm.js [hunk 1]
```diff
@@ -15,47 +14,69 @@ import { IncomeDeclaration } from "./IncomeDeclaration";
 const databaseURL = process.env.APP_URL  
 
 class EditRebateForm extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    const { rebateForm: { fields: {income} } } = this.props
+
+    const hasOtheIncome = income && income.other_income
```
Comment (jvdanker) [inline app/javascript/components/EditForm.js]: typo

File: app/javascript/components/EditForm.js [hunk 2]
```diff
@@ -15,47 +14,69 @@ import { IncomeDeclaration } from "./IncomeDeclaration";
 const databaseURL = process.env.APP_URL
```
Comment (jvdanker) [inline app/javascript/components/EditForm.js]: why did you name this 'databaseUrl'?

File: app/javascript/components/EditForm.js [hunk 3]
```diff
@@ -76,33 +94,43 @@ class EditRebateForm extends React.Component {
...
-                ? RadioInput(field)
-                : SingleInput(field)
+                ? RadioInput({...field, isEditable})
+                : SingleInput({...field, isEditable})
               })}
             </div>
             { values.fields.moved_within_rating_year == 'yes'
             ? <div className="flex-row">
                 {map(conditionalsFields, (field) => {
                 return field.type == 'radio'
```
Comment (jvdanker) [inline app/javascript/components/EditForm.js]: I prefer to use strict comparison

File: app/javascript/components/EditForm.js [hunk 4]
```diff
@@ -76,33 +94,43 @@ class EditRebateForm extends React.Component {
             <div className="flex-row">
               {map(customerDetailFields, (field) => {
                 return field.type == 'radio'
-                ? RadioInput(field)
-                : SingleInput(field)
+                ? RadioInput({...field, isEditable})
+                : SingleInput({...field, isEditable})
               })}
             </div>
             { values.fields.moved_within_rating_year == 'yes'
```
Comment (jvdanker) [inline app/javascript/components/EditForm.js]: For conditional rendering of components, I prefer to use `{ <condition> && <component /> }` over the ternary operator

File: app/javascript/components/EditForm.js [hunk 5]
```diff
@@ -15,47 +14,69 @@ import { IncomeDeclaration } from "./IncomeDeclaration";
...
+    const partnerKeys = hasOtheIncome ? Object.keys(income.other_income.partner) : []
+
+    const uniqKeys = uniq(applicantKeys.concat(partnerKeys))
+
+    this.state = { otherIncomeFields: uniqKeys }
+  }
+
+  addNewIncomeValue (values) {
+    this.setState({otherIncomeFields: this.state.otherIncomeFields.concat(values.newIncomeField)})
+    values.newIncomeField = null
```
Comment (jvdanker) [inline app/javascript/components/EditForm.js]: this isn't a 'pure' function and could cause side effects

File: app/javascript/components/EditForm.js [hunk 6]
```diff
@@ -76,33 +94,43 @@ class EditRebateForm extends React.Component {
...
+            <div className={'flex-row'}>
+              <Field
+                className='rebate-search-input flex-item'
+                name="newIncomeField"
+                component="input"
+                readOnly={!isEditable}
+              />
               <button
+                className='one-third rebate-add-income-button'
+                disabled={!isEditable || !(values.newIncomeField)}
```
Comment (jvdanker) [inline app/javascript/components/EditForm.js]: if this button should be enabled if it is `editable` *and* it is for a new incomefield, then I think writing it like
`! (isEditable && values.newIncomeField)` is easier to understand

File: app/javascript/components/inputs.js [hunk 7]
```diff
@@ -14,81 +14,51 @@ export function SingleInput ({ id, label, placeholder, fullWidth, type = "text"
       placeholder= {placeholder}
       component='input'
       type={type}
+      min={type == 'number' ? 0 : null}
+      readOnly={!isEditable}
```
Comment (jvdanker) [inline app/javascript/components/inputs.js]: just a general observation; you're using `isEditable` in every component to set the state of `readOnly`. In every occasion you have to negate it, so why don't you use `readOnly` instead?

File: app/javascript/components/EditForm.js (hunk 1 above)
Comment (mischa-s) [reply app/javascript/components/EditForm.js]: :laughing:

File: app/javascript/components/inputs.js (hunk 7 above)
Comment (mischa-s) [reply app/javascript/components/inputs.js]: yeah I didn't like it, was jamming it in at the end of the day, have changed it now

File: app/javascript/components/EditForm.js (hunk 6 above)
Comment (mischa-s) [reply app/javascript/components/EditForm.js]: having made changed `isEditable` to `isReadOnly` I find
`isReadOnly || !values.newIncomeField`
clearer then
`!(!isReadOnly && values.newIncomeField)`

File: app/javascript/components/EditForm.js (hunk 2 above)
Comment (mischa-s) [reply app/javascript/components/EditForm.js]: it is how I would normally do it for a separate react app but its true that in this case it is part of the same app, have changed it

**Pull request merged**